crates/matrix-sdk/src/room/joined.rs

```diff
@@ -644,64 +643,82 @@ impl Joined {
 
-            (response, file)
-        } else {
-            let response = self.client.upload(content_type, &mut reader).await?;
-            (response, None)
-        };
+            let res: Result<MessageType> = Ok(match content_type.type_() {
+                mime::IMAGE => {
+                    MessageType::Image(ImageMessageEventContent::encrypted(body.to_owned(), file))
+                }
+                mime::AUDIO => {
+                    MessageType::Audio(AudioMessageEventContent::encrypted(body.to_owned(), file))
+                }
+                mime::VIDEO => {
+                    MessageType::Video(VideoMessageEventContent::encrypted(body.to_owned(), file))
+                }
+                _ => MessageType::File(FileMessageEventContent::encrypted(body.to_owned(), file)),
+            });
+            res
+        }
 
-        let url = response.content_uri;
+        async fn get_content_unencrypted<R: Read>(
+            j: &Joined,
+            body: &str,
+            content_type: &Mime,
+            mut reader: &mut R,
+        ) -> Result<MessageType> {
+            let response = j.client.upload(content_type, &mut reader).await?;
 
-        let content = match content_type.type_() {
-            mime::IMAGE => {
-                // TODO create a thumbnail using the image crate?.
-                MessageType::Image(assign!(
-                    ImageMessageEventContent::plain(body.to_owned(), url, None),
-                    { file }
-                ))
-            }
-            mime::AUDIO => MessageType::Audio(assign!(
-                AudioMessageEventContent::plain(body.to_owned(), url, None),
-                { file }
-            )),
-            mime::VIDEO => MessageType::Video(assign!(
-                VideoMessageEventContent::plain(body.to_owned(), url, None),
-                { file }
-            )),
-            _ => MessageType::File(assign!(
-                FileMessageEventContent::plain(body.to_owned(), url, None),
-                { file }
-            )),
+            let url = response.content_uri;
+
+            Ok(match content_type.type_() {
+                mime::IMAGE => {
+                    // TODO create a thumbnail using the image crate?.
+                    MessageType::Image(ImageMessageEventContent::plain(body.to_owned(), url, None))
+                }
+                mime::AUDIO => {
+                    MessageType::Audio(AudioMessageEventContent::plain(body.to_owned(), url, None))
+                }
+                mime::VIDEO => {
+                    MessageType::Video(VideoMessageEventContent::plain(body.to_owned(), url, None))
+                }
+                _ => MessageType::File(FileMessageEventContent::plain(body.to_owned(), url, None)),
+            })
+        }
+
+        #[cfg(feature = "encryption")]
+        let content = if self.is_encrypted() {
+            get_content_encrypted(self, body, reader).await?
```

Comment:
It's a bit weird to have a free function that takes self here, let's make those two methods on the `Client` since they don't seem to need the room. 

The encrypted variant can be put into the `encryption` module which puts the method automatically behind the feature flag.

Reply:
Done. Although it seems like all methods in encryption/mod.rs still have a feature-cfg directive, so I've added one here as well.